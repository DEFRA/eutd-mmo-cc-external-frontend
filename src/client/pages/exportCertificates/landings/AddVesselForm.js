import React from 'react';
import { withTranslation } from 'react-i18next';
import { LabelText, ErrorText } from 'govuk-react';
import DateFieldWithPicker from '../../../components/DateFieldWithPicker';
import Details from '../../../components/elements/Details';
import GovukVesselsAutocomplete from '../../../components/GovukVesselsAutocomplete';
import moment from 'moment';
import { useTranslation } from 'react-i18next';
import { isEmpty } from 'lodash';

function AddVesselForm(props) {
  const faoAreas = [
    'FAO18',
    'FAO21',
    'FAO27',
    'FAO31',
    'FAO34',
    'FAO37',
    'FAO41',
    'FAO47',
    'FAO48',
    'FAO51',
    'FAO57',
    'FAO58',
    'FAO61',
    'FAO67',
    'FAO71',
    'FAO77',
    'FAO81',
    'FAO87',
    'FAO88',
  ];
  const {t} = useTranslation();

  const {
    dateLanded,
    onDateChange,
    dateLandedValue,
    faoArea,
    onFaoChange,
    vessel,
    vesselOptions,
    searchVessels,
    onVesselChange,
    clearSearchResults,
    errors = {}
  } = props;


  const splitErrorParams = (error) => {
    const isErrorWithParams = error.includes('-');
    const splitError = error.split('-');
    return isErrorWithParams ? t(splitError[0], {dynamicValue :splitError[1]}) : t(error);
  };


  const errorTranslate = (errs) => {
    let errorData = {};
    if(!isEmpty(errs)) {
      for (const key of Object.keys(errs)) {
        errorData[key] = splitErrorParams(errs[key]);
      }
    }
    return errorData;
  };

  const errorsData = errorTranslate(errors);

  let dateLandedRestrictionError;
  if (errors.dateLandedError) {
    let params = {};
    const dateLandedErrArray = errors.dateLandedError.split('-');
    if (dateLandedErrArray[0] == 'ccAddLandingDateLandedRestrictedError') {
      params = { product: dateLandedErrArray[1] };
    } else if (
      dateLandedErrArray[0] ==
      'ccUploadFilePageTableDateLandedFutureMaximumDaysError'
    ) {
      params = { dynamicValue: dateLandedErrArray[1] };
    }

    dateLandedRestrictionError = t(dateLandedErrArray[0], params);
  }


  const parsedDate = moment(dateLanded, ['YYYY-M-D', 'YYYY-MM-DD'], true);

  const vesselLandedDate = parsedDate.isValid() ? parsedDate : null;

  const setVessel = (term, item) => {
    if (term !== t('ccAddVesselFormVesselQueryPrompt') && term !== t('ccAddVesselFormVesselDateQueryPrompt')) {
      if (item && item.vesselName) {
        onVesselChange({ ...item });
      } else {
        onVesselChange({ label: term });
      }
    }
  };

  const getItems = (vessels) => {
    if (!vesselLandedDate) {
      vessels = [];
    }

    if (
      Array.isArray(vessels) &&
      vessels.length === 0 &&
      (!vessel.label || vessel.label.length < 2)
    ) {
      vessels.push({
        label: vesselLandedDate ? t('ccAddVesselFormVesselQueryPrompt') : t('ccAddVesselFormVesselDateQueryPrompt'),
      });
    }
    return vessels;
  };

  return (
    <div className="add-landings-form">
       <DateFieldWithPicker
        id='dateLanded'
        errors={dateLandedRestrictionError || errorsData.dateLandedError}
        onDateChange={onDateChange}
        dateFormat='YYYY-MM-DD'
        labelText={t('ccAddLandingDateLandedLabel')}
        labelTextClass='label-landings-form'
        date={dateLandedValue}
      />
      <LabelText className="label-landings-form">{t('ccAddLandingCatchAreaLabel')}</LabelText>
      {errorsData.faoAreaError && <ErrorText>{errorsData.faoAreaError}</ErrorText>}
      <select
        className={`autocomplete__input--default faoArea ${errorsData.faoAreaError ? 'autocomplete__error' : 'autocomplete__input'}`}
        id="select-faoArea"
        name="faoArea"
        value={faoArea}
        onChange={onFaoChange}
        style={{ backgroundColor: 'white' }}
      >
        {faoAreas.map((area) => (
          <option key={area} id={'faoArea_' + area} value={area}>
            {area}
          </option>
        ))}
      </select>
      <LabelText className="label-landings-form">{t('ccAddLandingVesselNameLabel')}</LabelText>
      <GovukVesselsAutocomplete
        onChange={setVessel}
        error={errorsData}
        hideErrorMessage={false}
        search={searchVessels}
        selectedItemName={vessel.label || ''}
        searchResults={getItems(vesselOptions || [])}
        getItem={(item) => item.label}
        clearSearchResults={clearSearchResults}
        controlName="vessel.vesselName"
        controlId="select-vessel"
        dateLanded={vesselLandedDate}
        hideBorder={true}
        translate={t}
      />
      <Details
        summary={t('ccAddLandingHelpSectionLinkText')}
        details={t('ccAddLandingHelpSectionContent')}
      />
    </div>
  );
}

export default withTranslation() (AddVesselForm);
