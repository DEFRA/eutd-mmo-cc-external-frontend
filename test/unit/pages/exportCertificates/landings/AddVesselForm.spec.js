import AddVesselForm from '../../../../../src/client/pages/exportCertificates/landings/AddVesselForm';
import { mount } from 'enzyme';
import React from 'react';
import moment from 'moment';
import { act } from 'react-dom/test-utils';
import { render } from '@testing-library/react';


describe('AddVesselForm', () => {
  let wrapper;

  const faoArea = 'FAO27';
  const vessel = { label: '' };
  const errors = {
    dateLandedError: 'Enter the date landed',
    'vessel.vesselNameError': 'Select a vessel from the list'
  };

  const dateLanded = moment('2021-12-01',['YYYY-M-D', 'YYYY-MM-DD'], true);
  const onDateChange = jest.fn();

  const onFaoChange = jest.fn();
  const vesselOptions = [];
  const searchVessels = jest.fn();
  const onVesselChange = jest.fn();
  const clearSearchResults = jest.fn();
  const getDateLanded = jest.fn();
  const isDateLandedComplete = jest.fn();

  beforeAll(() => {
    wrapper = mount(
      <AddVesselForm
        dateLanded={dateLanded}
        onDateChange={onDateChange}
        faoArea={faoArea}
        onFaoChange={onFaoChange}
        vessel={vessel}
        vesselOptions={vesselOptions}
        searchVessels={searchVessels}
        onVesselChange={onVesselChange}
        clearSearchResults={clearSearchResults}
        getDateLanded = {getDateLanded}
        isDateLandedComplete = {isDateLandedComplete}
        errors={errors}
      />
    );
  });

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('will render', () => {
    expect(wrapper).not.toBeNull();
    expect(wrapper.text()).toContain('Date Landed');
  });

  it('should take a snapshot of the whole page', () => {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  });

  it('will set the date from the props', () => {
    expect(wrapper.find('AddVesselForm').prop('dateLanded')).toEqual(
      dateLanded
    );
  });

  it('will set the FAO area from the props', () => {
    expect(wrapper.find('#select-faoArea').prop('value')).toBe(faoArea);
  });

  it('will handle the FAO area changing', () => {
    wrapper
      .find('#select-faoArea')
      .simulate('change', { target: { value: 'FAO31' } });

    expect(onFaoChange).toHaveBeenCalledWith(
      expect.objectContaining({ target: { value: 'FAO31' } })
    );
  });

  it('will handle searching', () => {
    const search = wrapper.find('GovukVesselsAutocomplete').prop('search');

    search();

    expect(searchVessels).toHaveBeenCalled();
  });

  it('will handle clearing search results', () => {
    const clear = wrapper
      .find('GovukVesselsAutocomplete')
      .prop('clearSearchResults');

    clear();

    expect(clearSearchResults).toHaveBeenCalled();
  });

  it('will handle typing in a vessel name', () => {
    const onChange = wrapper.find('GovukVesselsAutocomplete').prop('onChange');

    onChange('wiron 5', null);

    expect(onVesselChange).toHaveBeenCalledWith({ label: 'wiron 5' });
  });

  it('will handle finding a vessel', () => {
    const onChange = wrapper.find('GovukVesselsAutocomplete').prop('onChange');

    onChange('wiron 5', { vesselName: 'WIRON 5' });

    expect(onVesselChange).toHaveBeenCalledWith({ vesselName: 'WIRON 5' });
  });

  it('should not allow user to select a vessel when date has not been set', () => {
    const ac = wrapper.find('GovukVesselsAutocomplete');
    expect(ac.exists()).toBeTruthy();
    act(() => ac.props().onChange('Enter a valid date landed to enable Vessel name'));

     expect(wrapper.find('AddVesselForm').props().vessel).toEqual(
      {
        label: ''
      }
    );
  });

  it('will parse vessel data for the autocomplete component', () => {
    const getItem = wrapper.find('GovukVesselsAutocomplete').prop('getItem');

    expect(getItem({ label: 'WIRON 5', pln: '5500' })).toEqual('WIRON 5');
  });

  it('will parse the date input field to a moment object for the autocomplete component', () => {
    const autoCompleteDate = wrapper
      .find('GovukVesselsAutocomplete')
      .prop('dateLanded');

    const expected = moment(dateLanded, 'YYYY-M-D');

    expect(autoCompleteDate.isSame(expected));
  });

  it('will pass null to the autocomplete component if the date does not match the correct format', () => {
    const invalidDateWrapper = mount(
      <AddVesselForm
        dateLanded={'20-05-14'}
        onDateChange={onDateChange}
        faoArea={faoArea}
        onFaoChange={onFaoChange}
        vessel={vessel}
        vesselOptions={vesselOptions}
        searchVessels={searchVessels}
        onVesselChange={onVesselChange}
        clearSearchResults={clearSearchResults}
        getDateLanded = {getDateLanded}
        isDateLandedComplete = {isDateLandedComplete}
        errors={errors}
      />
    );
    const autoCompleteDate = invalidDateWrapper
      .find('GovukVesselsAutocomplete')
      .prop('dateLanded');

    expect(autoCompleteDate).toBeNull();
  });

  it('will display validation errors for each component', () => {
    expect(wrapper.find('DateField').at(0).text()).toContain(errors.dateLandedError);
    expect(wrapper.find('GovukVesselsAutocomplete').at(0).text()).toContain(errors['vessel.vesselNameError']);
  });

  it('will display an error on the FAO areas field', () => {
    const invalidFAOError = {
      faoAreaError: 'Catch area is missing'
    };

    const invalidFAOWrapper = mount(
      <AddVesselForm
        dateLanded={'20-05-14'}
        onDateChange={onDateChange}
        faoArea={faoArea}
        onFaoChange={onFaoChange}
        vessel={vessel}
        vesselOptions={vesselOptions}
        searchVessels={searchVessels}
        onVesselChange={onVesselChange}
        clearSearchResults={clearSearchResults}
        getDateLanded = {getDateLanded}
        isDateLandedComplete = {isDateLandedComplete}
        errors={invalidFAOError}
      />
    );

    expect(invalidFAOWrapper
      .find('#select-faoArea')
      .prop('className')).toBe('autocomplete__input--default faoArea autocomplete__error');
  });
});
