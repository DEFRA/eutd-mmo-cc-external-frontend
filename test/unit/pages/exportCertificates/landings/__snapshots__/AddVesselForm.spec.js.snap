// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddVesselForm should take a snapshot of the whole page 1`] = `
<div>
  <div
    class="add-landings-form"
  >
    <div
      class="date-field-with-picker"
      id="dateLanded"
    >
      <div
        class="column-field-picker"
      >
        <div
          class="date-field e1vcssft0 css-1bm99do-StyledContainer-StyledHoc e4u1u2g0"
        >
          <span
            class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
          >
            <span
              class="label-landings-form e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
              id="date-field-label-text"
            >
              Date Landed
            </span>
          </span>
          <span
            class="e1vcssft0 css-1whl0n4-StyledHint-StyledHoc e4u1u2g0"
          >
            For example, 31 03 1980
          </span>
          <span
            class="e1vcssft0 css-9vo93-StyledError-StyledHoc e4u1u2g0"
          >
            Enter the date landed
          </span>
          <div
            class="css-bw34j1-StyledList e107rt5v2"
          >
            <label
              class="e107rt5v1 css-iwpl6b-StyledLabel-StyledHoc-StyledLabel e4u1u2g0"
            >
              <span
                class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
              >
                Day
              </span>
              <input
                class="css-oy9rc0-StyledInput e107rt5v0"
                name="dayInputName"
                type="number"
                value=""
              />
            </label>
            <label
              class="e107rt5v1 css-iwpl6b-StyledLabel-StyledHoc-StyledLabel e4u1u2g0"
            >
              <span
                class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
              >
                Month
              </span>
              <input
                class="css-oy9rc0-StyledInput e107rt5v0"
                name="monthInputName"
                type="number"
                value=""
              />
            </label>
            <label
              class="e107rt5v1 css-17bpbza-StyledLabel-StyledHoc-StyledLabel e4u1u2g0"
            >
              <span
                class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
              >
                Year
              </span>
              <input
                class="css-oy9rc0-StyledInput e107rt5v0"
                name="yearInputName"
                type="number"
                value=""
              />
            </label>
          </div>
        </div>
        <div
          class="date-picker error"
        >
          <button>
            <img
              alt="calendar"
              src="/static/assets/images/calendar.png"
            />
          </button>
        </div>
      </div>
    </div>
    <span
      class="label-landings-form e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
    >
      Catch area
    </span>
    <select
      class="autocomplete__input--default faoArea autocomplete__input"
      id="select-faoArea"
      name="faoArea"
      style="background-color: white;"
    >
      <option
        id="faoArea_FAO18"
        value="FAO18"
      >
        FAO18
      </option>
      <option
        id="faoArea_FAO21"
        value="FAO21"
      >
        FAO21
      </option>
      <option
        id="faoArea_FAO27"
        value="FAO27"
      >
        FAO27
      </option>
      <option
        id="faoArea_FAO31"
        value="FAO31"
      >
        FAO31
      </option>
      <option
        id="faoArea_FAO34"
        value="FAO34"
      >
        FAO34
      </option>
      <option
        id="faoArea_FAO37"
        value="FAO37"
      >
        FAO37
      </option>
      <option
        id="faoArea_FAO41"
        value="FAO41"
      >
        FAO41
      </option>
      <option
        id="faoArea_FAO47"
        value="FAO47"
      >
        FAO47
      </option>
      <option
        id="faoArea_FAO48"
        value="FAO48"
      >
        FAO48
      </option>
      <option
        id="faoArea_FAO51"
        value="FAO51"
      >
        FAO51
      </option>
      <option
        id="faoArea_FAO57"
        value="FAO57"
      >
        FAO57
      </option>
      <option
        id="faoArea_FAO58"
        value="FAO58"
      >
        FAO58
      </option>
      <option
        id="faoArea_FAO61"
        value="FAO61"
      >
        FAO61
      </option>
      <option
        id="faoArea_FAO67"
        value="FAO67"
      >
        FAO67
      </option>
      <option
        id="faoArea_FAO71"
        value="FAO71"
      >
        FAO71
      </option>
      <option
        id="faoArea_FAO77"
        value="FAO77"
      >
        FAO77
      </option>
      <option
        id="faoArea_FAO81"
        value="FAO81"
      >
        FAO81
      </option>
      <option
        id="faoArea_FAO87"
        value="FAO87"
      >
        FAO87
      </option>
      <option
        id="faoArea_FAO88"
        value="FAO88"
      >
        FAO88
      </option>
    </select>
    <span
      class="label-landings-form e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
    >
      Vessel name or PLN
    </span>
    <label
      class="e1vcssft0 css-137pbp-StyledLabel-StyledHoc e4u1u2g0"
      for="select-vessel"
    >
      <span
        class="e1vcssft0 css-9vo93-StyledError-StyledHoc e4u1u2g0"
      >
        Select a vessel from the list
      </span>
      <div
        style="position: relative;"
      >
        <div
          style="border: 1em;"
        >
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            autocomplete="off"
            class="css-1osip2e"
            id="select-vessel"
            name="vessel.vesselName"
            role="combobox"
            style="border: 4px solid #b10e1e;"
            value=""
          />
        </div>
      </div>
    </label>
    <details>
      <summary
        class="govuk-details__summary"
      >
        I cannot find the vessel
      </summary>
      <div
        class="govuk-details__text "
      >
        If the vessel you need cannot be found, search for and select 'Vessel not found (N/A)'.
      </div>
    </details>
  </div>
</div>
`;
